# RETHINKING the benchmarking of taxonomic profilers for shotugun metagenomics data

***
## Introduction
Microbial identification and accurate abundance estimation are still challenging in the metagenomics analysis yet poorly understood and evaluated using appropriate metrics. 

**Microbial identification: false positive assignment** High FP rate in the microbial identification is among the most critical yet-to-resolve issues of metagenomics (Ye et al., 2019). This problem leads to severe over-estimation of species richness, confounds true signals in the construction of taxonomic and functional profiles, and impairs the efficancy of compositionality-based community analysis.

**Abundance estimation: sequence abundance VS taxonomic abundance** Microbial abundance in the metagenomics data can be considered either as the relative abundance of reads from each taxa ("sequence abundance") or by inferring abundance of the number of individuals from each taxa by correcting read counts for genome size ("taxonomic abundance"). However, such correction for genome length into abundance estimation is usually missed in most state-of-the-art profilers. Even though it can be manually performed by reweighting the read counts after classification, [previous benchmarking studies](https://www.sciencedirect.com/science/article/pii/S0092867419307755) did not perform appropriate corrections and still used "raw" abundance profiles generated from benchmarked profilers for performance comparisons , resulting in highly misleading or even contradictory conclusions.

## Previous benchmarking results
In the [recent paper](https://www.sciencedirect.com/science/article/pii/S0092867419307755) published on CELL, authors review current state-of-the-art approaches for metagenomic analysis and evaluate the performance of 20 metagenomic classifiers by two key metrics (AUPR score and L2 distance) on simulated and experimental datasets. 
However, both AUPR score using actual abudance and L2 distance using 'raw' profiling results from profilers can provide misleading benchmarking results.

## RE-benchmarking using differential metrics
### Data simulation using a series of genomes (N=10, 20, 50, 200, and 500) with explicit genome length.
We employed [Wgsim](https://github.com/lh3/wgsim) to simulate metagenomics data with a given number of genomes.
Wgsim is originally a small tool for simulating sequence reads from a reference genome. Here we mixed the genomic sequences from multiple microbial taxa at certain percentages in a fasta file and then took it as input for Wgsim.


### AUPR score alone is not able to present the performance of a metagenomics classifier/profiler in the microbial identification
![AUPR](NewSim_auprc_bracken-metaplhan2_boxplot.png)
The AUPR score can be calculated using either actual abundance or "binary" abundance. In the [CELL paper](https://www.sciencedirect.com/science/article/pii/S0092867419307755), authors used AUPR which essentially calculated as the area under the precision-recall curve by ranging the actual abundance threshold from 0 to 1.0. However, this AUPR score biased toward low-precision and high-recall classifiers (such as Bracken or Kracken with a more comprehensive database), which canâ€™t reflect the high-false-positive issue. In the real-world, it is impractical for a user to arbitrarily set an abundance threshold to filter potential false positives (especially for a complex microbiome sample) as the abundance of TP and FP are often very close to each other. Therefore, other than AUPR score, we here proposed to calculate more metrics to reflect the relative performance of a given metagenomics tool on microbial identification, especially the false-positives issue in the observed profiles. 
### "Taxonomic abundance" was usually confounded with "sequence abundance" in the previous benchmarking works.
![L2_dist](NewSim_L2.dist_bracken-metaplhan2_boxplot.png)
In the previous benchmarking study, the L2 distance was calculated between observed and expected abundance profiles. Notably, abundance here can be calculated either as the relative abundance of sequencing reads from microbial taxa or the relative abundance of individuals from a given microbial taxa. In the [CELL paper](https://www.sciencedirect.com/science/article/pii/S0092867419307755), the ground-truth abundance was sequence abundance, whereas a raw abundance profile generated by benchmarked metagenomics tools can be either sequence abundance (such as Bracken, Kracken) or taxonomic abundance (MetaPhlAn2). This inconsistence in the output abudance profiles across metagenomics classifiers largely affect or mislead the benchmarking results, where Bracken typically outperform others.
Here, we rebenchmarked the metagenomics tools on the simulation datasets with known genome size so that we can compare the performance of abundance estimation across metagenomics tools under the context of either sequence abundance and taxonomic abundance. We found that if the sequence abundance was used as ground truth to compare the performance of these two tools, Bracken outperforms MetaPhlAn2, however if taxonomic abundance was used as ground truth, MetaPhlAn2 is better than Bracken (Figure as above). Therefore, the normalization of raw abudance profiles from metagnomics may faciliate a more comprehensive benchmarking.

## Reference
* Ye, S.H., Siddle, K.J., Park, D.J., and Sabeti, P.C. (2019). Benchmarking Metagenomics Tools for Taxonomic Classification. Cell 178, 779-794.
* D., Foox, J., Ahsanuddin, S., et al. (2017). Comprehensive benchmarking and ensemble approaches for metagenomic classifiers. Genome Biology 18.
* Segata, N., Waldron, L., Ballarini, A., Narasimhan, V., Jousson, O., and Huttenhower, C. (2012). Metagenomic microbial community profiling using unique clade-specific marker genes. Nat Methods 9, 811-+.
* Mende, D.R., Waller, A.S., Sunagawa, S., Jarvelin, A.I., Chan, M.M., Arumugam, M., Raes, J., and Bork, P. (2012). Assessment of Metagenomic Assembly Using Simulated Next Generation Sequencing Data. Plos One 7.
* Truong, D.T., Franzosa, E.A., Tickle, T.L., Scholz, M., Weingart, G., Pasolli, E., Tett, A., Huttenhower, C., and Segata, N. (2015). MetaPhlAn2 for enhanced metagenomic taxonomic profiling. Nat Methods 12, 902-903.
* Wood, D.E., and Salzberg, S.L. (2014). Kraken: ultrafast metagenomic sequence classification using exact alignments. Genome Biology 15.



